File: goblobBase/src/main/java/com/cauchymop/goblob/model/GoGameController.java
package com.cauchymop.goblob.model;

import com.cauchymop.goblob.proto.PlayGameData;
import com.cauchymop.goblob.proto.PlayGameData.GameData.Phase;
import com.google.common.base.Preconditions;
import com.google.common.collect.Sets;

import java.io.Serializable;
import java.util.List;

import static com.cauchymop.goblob.proto.PlayGameData.Color;
import static com.cauchymop.goblob.proto.PlayGameData.GameConfiguration;
import static com.cauchymop.goblob.proto.PlayGameData.GameData;
import static com.cauchymop.goblob.proto.PlayGameData.GoPlayer;
import static com.cauchymop.goblob.proto.PlayGameData.MatchEndStatus;
import static com.cauchymop.goblob.proto.PlayGameData.Move;
import static com.cauchymop.goblob.proto.PlayGameData.Position;
import static com.cauchymop.goblob.proto.PlayGameData.Score;

/**
 * Class to handle interactions between the players and the {@link GoGame}.
 */
public class GoGameController implements Serializable {

  transient GameDatas gameDatas;

  private GoGame goGame;
  private GameData.Builder gameData;
  private GameData initialGameData;
  private Analytics analytics;

  public GoGameController(GameDatas gameDatas, GameData gameData, Analytics analytics) {
    this.gameDatas = gameDatas;
    this.initialGameData = gameData;
    this.gameData = Preconditions.checkNotNull(gameData).toBuilder();
    this.analytics = analytics;
    // The GoGame settings can change during the configuration, so we postpone its creation.
    if (gameData.getPhase() != Phase.CONFIGURATION) {
      createGoGame();
      for (Move move : this.gameData.getMoveList()) {
        goGame.play(getPos(move));
      }
    }
  }

  public void createGoGame() {
    GameConfiguration gameConfiguration = getGameConfiguration();
    goGame = new GoGame(gameConfiguration.getBoardSize(), gameConfiguration.getHandicap());
  }

  public Score getScore() {
    return getMatchEndStatus().getScore();
  }

  public boolean undo() {
    if (canUndo()) {
      gameData.addRedo(0, removeLastMove());
      goGame.undo();
      return true;
    }
    return false;
  }

  public boolean redo() {
    if (canRedo()) {
      playMove(gameData.getRedo(0));
      return true;
    }
    return false;
  }

  private boolean playMove(Move move) {
    if (gameData.getPhase() == Phase.IN_GAME && goGame.play(getPos(move))) {
      updateRedoForMove(move);
      gameData.addMove(move);
      gameData.setTurn(getOpponentColor());
      checkForMatchEnd();
      analytics.movePlayed(getGameConfiguration(), move);
      return true;
    } else {
      analytics.invalidMovePlayed(getGameConfiguration());
      return false;
    }
  }

  public GoPlayer getCurrentPlayer() {
    return gameDatas.getCurrentPlayer(gameData);
  }

  public GoPlayer getOpponent() {
    return gameDatas.getGoPlayer(gameData, getOpponentColor());
  }

  public Color getCurrentColor() {
    return gameDatas.getCurrentColor(gameData);
  }

  @Override
  public String toString() {
    return String.format("GoGameController(GoGame=%s, black=%s, white=%s, end=%s)",
        goGame, getBlackPlayer(), getWhitePlayer(), getMatchEndStatus());
  }

  public GoGame getGame() {
    return goGame;
  }

  public GameData buildGameData() {
    gameData.setSequenceNumber(gameData.getSequenceNumber() + 1);
    return gameData.build();
  }

  public GameConfiguration getGameConfiguration() {
    return gameData.getGameConfiguration();
  }

  private Move removeLastMove() {
    int lastIndex = gameData.getMoveCount() - 1;
    Move lastMove = gameData.getMove(lastIndex);
    gameData.removeMove(lastIndex);
    return lastMove;
  }

  public boolean isLocalTurn() {
    return gameDatas.isLocalTurn(gameData);
  }

  private boolean toggleDeadStone(Position position) {
    if (goGame.getColor(position.getX(), position.getY()) == null) {
      return false;
    }
    int index = getMatchEndStatus().getDeadStoneList().indexOf(position);
    MatchEndStatus.Builder matchEndStatus = gameData.getMatchEndStatusBuilder();
    if (index == -1) {
      matchEndStatus.addDeadStone(position);
    } else {
      matchEndStatus.removeDeadStone(index);
    }
    matchEndStatus.setLastModifier(gameData.getTurn());
    matchEndStatus.setScore(calculateScore());
    analytics.deadStoneToggled(getGameConfiguration());
    return true;
  }

  public List<PlayGameData.Position> getDeadStones() {
    return getMatchEndStatus().getDeadStoneList();
  }

  public void markingTurnDone() {
    if (isLocalGame() || !getMatchEndStatus().getLastModifier().equals(getCurrentColor())) {
      gameData.setPhase(Phase.FINISHED);
      analytics.gameFinished(getGameConfiguration(), getScore());
    }
    gameData.setTurn(getOpponentColor());
  }

  public boolean canUndo() {
    return isLocalGame() && gameData.getPhase() == Phase.IN_GAME && !gameData.getMoveList().isEmpty();
  }

  public boolean canRedo() {
    return isLocalGame() && gameData.getPhase() == Phase.IN_GAME && !gameData.getRedoList().isEmpty();
  }

  public void resign() {
    gameData.setPhase(Phase.FINISHED);
    Score.Builder score = gameData.getMatchEndStatusBuilder().getScoreBuilder();
    score.setWinner(getOpponentColor());
    score.setResigned(true);
    analytics.gameFinished(getGameConfiguration(), getScore());
  }

  public boolean isLocalGame() {
    return gameDatas.isLocalGame(gameData);
  }

  private Color getOpponentColor() {
    return GoBoard.getOpponent(getCurrentColor());
  }

  private Score calculateScore() {
    ScoreGenerator scoreGenerator = new ScoreGenerator(goGame.getBoard(),
        Sets.newHashSet(getDeadStones()), getGameConfiguration().getKomi());
    return scoreGenerator.getScore();
  }

  private void updateRedoForMove(Move move) {
    if (gameData.getRedoCount() == 0) {
      return;
    }
    if (move.equals(gameData.getRedo(0))) {
      gameData.removeRedo(0);
    } else {
      gameData.clearRedo();
    }
  }

  private void checkForMatchEnd() {
    if (goGame.isGameEnd()) {
      gameData.setPhase(Phase.DEAD_STONE_MARKING);
      Color lastModifier = GoBoard.getOpponent(goGame.getCurrentColor());
      gameData.getMatchEndStatusBuilder()
          .setLastModifier(lastModifier)
          .setScore(calculateScore());
    }
  }

  private int getPos(Move move) {
    switch (move.getType()) {
      case MOVE:
        Position position = move.getPosition();
        return goGame.getPos(position.getX(), position.getY());
      case PASS:
        return goGame.getPassValue();
      default:
        throw new RuntimeException("Invalid Move");
    }
  }

  private MatchEndStatus getMatchEndStatus() {
    return gameData.getMatchEndStatus();
  }

  private GoPlayer getWhitePlayer() {
    return getGameConfiguration().getWhite();
  }

  private GoPlayer getBlackPlayer() {
    return getGameConfiguration().getBlack();
  }

  public boolean isGameFinished() {
    return gameData.getPhase() == Phase.FINISHED;
  }

  public Phase getPhase() {
    return gameData.getPhase();
  }

  public boolean playMoveOrToggleDeadStone(Move move) {
    switch(getPhase()) {
      case IN_GAME:
        return playMove(move);
      case DEAD_STONE_MARKING:
        return toggleDeadStone(move.getPosition());
      default:
        throw new RuntimeException("Invalid mode");
    }
  }

  public GoPlayer getWinner() {
    return gameDatas.getGoPlayer(gameData, gameData.getMatchEndStatus().getScore().getWinner());
  }

  public void validateConfiguration() {

    if (isConfigurationAgreed(initialGameData, getGameConfiguration())) {
      gameData.setPhase(Phase.IN_GAME);
      createGoGame();
    } else {
      gameData.setPhase(Phase.CONFIGURATION);
    }
    gameData.setTurn(computeConfigurationNextTurn());
    analytics.configurationChanged(gameData);
  }

  private boolean isConfigurationAgreed(GameData initialGame,
      GameConfiguration newGameConfiguration) {
    return initialGame.getGameConfiguration().getGameType() == PlayGameData.GameType.LOCAL
        || (initialGame.getPhase() == Phase.CONFIGURATION
          && initialGame.getGameConfiguration().equals(newGameConfiguration));
  }

  private PlayGameData.Color computeConfigurationNextTurn() {
    if (getPhase() == Phase.CONFIGURATION) {
      return gameDatas.getOpponentColor(getGameConfiguration());
    } else if (getPhase() == Phase.IN_GAME) {
      return gameDatas.computeInGameTurn(getGameConfiguration(), 0);
    } else {
      throw new IllegalArgumentException("Invalid phase: " + getPhase());
    }
  }

  public GoPlayer getPlayerForColor(Color player) {
    return gameDatas.getGoPlayer(gameData, player);
  }

  public String getMatchId() {
    return gameData.getMatchId();
  }

  public void setBlackPlayerName(String blackPlayerName) {
    gameData.getGameConfigurationBuilder().getBlackBuilder().setName(blackPlayerName);
  }

  public void setWhitePlayerName(String whitePlayerName) {
    gameData.getGameConfigurationBuilder().getWhiteBuilder().setName(whitePlayerName);
  }

  public void setBoardSize(int boardSize) {
    gameData.getGameConfigurationBuilder().setBoardSize(boardSize);
  }

  public void setKomi(float komi) {
    gameData.getGameConfigurationBuilder().setKomi(komi);
  }

  public void setHandicap(int handicap) {
    gameData.getGameConfigurationBuilder().setHandicap(handicap);
  }

  public void swapPlayers() {
    GameConfiguration gameConfiguration = getGameConfiguration();
    PlayGameData.GoPlayer black = gameConfiguration.getBlack();
    PlayGameData.GoPlayer white = gameConfiguration.getWhite();
    gameData.getGameConfigurationBuilder().setBlack(white);
    gameData.getGameConfigurationBuilder().setWhite(black);
    gameData.setTurn(getOpponentColor());
  }

  public boolean pass() {
    return playMove(gameDatas.createPassMove());
  }
}

File: goblobBase/src/main/java/com/cauchymop/goblob/model/GoGameController.kt
package com.cauchymop.goblob.model

import com.cauchymop.goblob.proto.PlayGameData
import com.cauchymop.goblob.proto.PlayGameData.*
import com.cauchymop.goblob.proto.PlayGameData.GameData.Phase
import com.cauchymop.goblob.proto.PlayGameData.GameData.Phase.*
import com.cauchymop.goblob.proto.PlayGameData.GameType.LOCAL
import com.google.common.base.Preconditions
import com.google.common.collect.Sets
import java.io.Serializable
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.properties.Delegates

/**
 * Class to handle interactions between the players and the [GoGame].
 */
@Singleton
class GoGameController @Inject constructor(
    @field:Transient private var gameDatas: GameDatas,
    private val analytics: Analytics) : Serializable {

  var game: GoGame? = null
    private set

  private var gameDataBuilder: GameData.Builder by Delegates.notNull()

  private var initialGameData: GameData by Delegates.notNull()

  var gameData: GameData
    get() = gameDataBuilder.apply { sequenceNumber++ }.build()
    set(value) {
      this.initialGameData = value
      this.gameDataBuilder = Preconditions.checkNotNull(value).toBuilder()
      // The GoGame settings can change during the configuration, so we postpone its creation.
      if (value.phase != CONFIGURATION) {
        createGoGame()
        for (move in this.gameDataBuilder.moveList) {
          game!!.play(getPos(move))
        }
      }
    }

  val score: Score
    get() = matchEndStatus.score

  val currentPlayer: GoPlayer
    get() = gameDatas.getCurrentPlayer(gameDataBuilder)

  val opponent: GoPlayer
    get() = gameDatas.getGoPlayer(gameDataBuilder, opponentColor)

  val currentColor: Color
    get() = gameDatas.getCurrentColor(gameDataBuilder)

  val gameConfiguration: GameConfiguration
    get() = gameDataBuilder.gameConfiguration

  val isLocalTurn: Boolean
    get() = gameDatas.isLocalTurn(gameDataBuilder)

  val deadStones: List<PlayGameData.Position>
    get() = matchEndStatus.deadStoneList

  val isLocalGame: Boolean
    get() = gameDatas.isLocalGame(gameDataBuilder)

  private val opponentColor: Color
    get() = GoBoard.getOpponent(currentColor)

  private val matchEndStatus: MatchEndStatus
    get() = gameDataBuilder.matchEndStatus

  private val whitePlayer: GoPlayer
    get() = gameConfiguration.white

  private val blackPlayer: GoPlayer
    get() = gameConfiguration.black

  val isGameFinished: Boolean
    get() = gameDataBuilder.phase == FINISHED

  val phase: Phase
    get() = gameDataBuilder.phase

  val winner: GoPlayer
    get() = gameDatas.getGoPlayer(gameDataBuilder, gameDataBuilder.matchEndStatus.score.winner)

  val matchId: String
    get() = gameDataBuilder.matchId

  fun undo(): Boolean {
    if (canUndo()) {
      gameDataBuilder.addRedo(0, removeLastMove())
      game!!.undo()
      return true
    }
    return false
  }

  fun redo(): Boolean {
    if (canRedo()) {
      playMove(gameDataBuilder.getRedo(0))
      return true
    }
    return false
  }

  override fun toString(): String {
    return String.format("GoGameController(GoGame=%s, black=%s, white=%s, end=%s)",
        game, blackPlayer, whitePlayer, matchEndStatus)
  }

  fun markingTurnDone() {
    if (isLocalGame || matchEndStatus.lastModifier != currentColor) {
      gameDataBuilder.phase = FINISHED
      analytics.gameFinished(gameConfiguration, score)
    }
    gameDataBuilder.turn = opponentColor
  }

  fun canUndo(): Boolean {
    return isLocalGame && gameDataBuilder.phase == IN_GAME && !gameDataBuilder.moveList.isEmpty()
  }

  fun canRedo(): Boolean {
    return isLocalGame && gameDataBuilder.phase == IN_GAME && !gameDataBuilder.redoList.isEmpty()
  }

  fun resign() {
    gameDataBuilder.phase = FINISHED
    gameDataBuilder.matchEndStatusBuilder.scoreBuilder.apply {
      winner = opponentColor
      resigned = true
    }
    analytics.gameFinished(gameConfiguration, score)
  }

  fun playMoveOrToggleDeadStone(move: Move): Boolean {
    when (phase) {
      IN_GAME -> return playMove(move)
      DEAD_STONE_MARKING -> return toggleDeadStone(move.position)
      else -> throw RuntimeException("Invalid mode")
    }
  }

  fun validateConfiguration() {

    if (isConfigurationAgreed()) {
      gameDataBuilder.phase = IN_GAME
      createGoGame()
    } else {
      gameDataBuilder.phase = CONFIGURATION
    }
    gameDataBuilder.turn = computeConfigurationNextTurn()
    analytics.configurationChanged(gameDataBuilder)
  }

  fun getPlayerForColor(player: Color): GoPlayer {
    return gameDatas.getGoPlayer(gameDataBuilder, player)
  }

  fun setBlackPlayerName(blackPlayerName: String) {
    gameDataBuilder.gameConfigurationBuilder.blackBuilder.name = blackPlayerName
  }

  fun setWhitePlayerName(whitePlayerName: String) {
    gameDataBuilder.gameConfigurationBuilder.whiteBuilder.name = whitePlayerName
  }

  fun setBoardSize(boardSize: Int) {
    gameDataBuilder.gameConfigurationBuilder.boardSize = boardSize
  }

  fun setKomi(komi: Float) {
    gameDataBuilder.gameConfigurationBuilder.komi = komi
  }

  fun setHandicap(handicap: Int) {
    gameDataBuilder.gameConfigurationBuilder.handicap = handicap
  }

  fun swapPlayers() {
    val gameConfiguration = gameConfiguration
    val black = gameConfiguration.black
    val white = gameConfiguration.white
    gameDataBuilder.gameConfigurationBuilder.black = white
    gameDataBuilder.gameConfigurationBuilder.white = black
    gameDataBuilder.turn = opponentColor
  }

  fun pass(): Boolean {
    return playMove(gameDatas.createPassMove())
  }

  private fun createGoGame() {
    game = GoGame(gameConfiguration.boardSize, gameConfiguration.handicap)
  }

  private fun playMove(move: Move): Boolean {
    if (gameDataBuilder.phase == IN_GAME && game!!.play(getPos(move))) {
      updateRedoForMove(move)
      gameDataBuilder.addMove(move)
      gameDataBuilder.turn = opponentColor
      checkForMatchEnd()
      analytics.movePlayed(gameConfiguration, move)
      return true
    } else {
      analytics.invalidMovePlayed(gameConfiguration)
      return false
    }
  }

  private fun removeLastMove(): Move {
    val lastIndex = gameDataBuilder.moveCount - 1
    val lastMove = gameDataBuilder.getMove(lastIndex)
    gameDataBuilder.removeMove(lastIndex)
    return lastMove
  }

  private fun toggleDeadStone(position: Position): Boolean {
    if (game!!.getColor(position.x, position.y) == null) {
      return false
    }
    val index = matchEndStatus.deadStoneList.indexOf(position)
    val matchEndStatus = gameDataBuilder.matchEndStatusBuilder
    if (index == -1) {
      matchEndStatus.addDeadStone(position)
    } else {
      matchEndStatus.removeDeadStone(index)
    }
    matchEndStatus.lastModifier = gameDataBuilder.turn
    matchEndStatus.score = calculateScore()
    analytics.deadStoneToggled(gameConfiguration)
    return true
  }

  private fun isConfigurationAgreed(): Boolean =
      initialGameData.gameConfiguration.gameType == LOCAL ||
          (initialGameData.phase == CONFIGURATION && initialGameData.gameConfiguration == gameConfiguration)

  private fun computeConfigurationNextTurn(): PlayGameData.Color = when (phase) {
    CONFIGURATION -> gameDatas.getOpponentColor(gameConfiguration)
    IN_GAME -> gameDatas.computeInGameTurn(gameConfiguration, 0)
    else -> throw IllegalArgumentException("Invalid phase: $phase")
  }

  private fun calculateScore(): Score {
    val scoreGenerator = ScoreGenerator(game!!.board,
        Sets.newHashSet(deadStones), gameConfiguration.komi)
    return scoreGenerator.score
  }

  private fun updateRedoForMove(move: Move) {
    if (gameDataBuilder.redoCount == 0) {
      return
    }
    if (move == gameDataBuilder.getRedo(0)) {
      gameDataBuilder.removeRedo(0)
    } else {
      gameDataBuilder.clearRedo()
    }
  }

  private fun checkForMatchEnd() {
    if (game!!.isGameEnd) {
      gameDataBuilder.phase = DEAD_STONE_MARKING
      val lastModifier = GoBoard.getOpponent(game!!.currentColor)
      gameDataBuilder.matchEndStatusBuilder
          .setLastModifier(lastModifier).score = calculateScore()
    }
  }

  private fun getPos(move: Move): Int {
    when (move.type) {
      PlayGameData.Move.MoveType.MOVE -> {
        val position = move.position
        return game!!.getPos(position.x, position.y)
      }
      PlayGameData.Move.MoveType.PASS -> return game!!.passValue
      else -> throw RuntimeException("Invalid Move")
    }
  }
}
